src/foam/classloader/ClassLoader.js

```diff
@@ -67,14 +67,25 @@ foam.CLASS({
     {
       name: 'arequire',
       class: 'foam.core.ContextMethod',
-      code: function(X, modelId) {
-        if (X.isRegistered(modelId)) return Promise.resolve();
-        if (this.pending[modelId]) return this.pending[modelId];
+      code: function(X, modelId, deps) {
+        // Contains models that depend on the modelId and have already been
+        // arequired. Used to avoid circular dependencies from waiting on
+        // eachother.
+        deps = deps || {};
+
+        if ( X.isRegistered(modelId) ) return Promise.resolve();
+        if ( deps[modelId] ) return Promise.resolve();
+        if ( this.pending[modelId] ) return this.pending[modelId];
+        deps[modelId] = true;
 
         var modelDao = X[foam.String.daoize(foam.core.Model.name)];
-        this.pending[modelId] = modelDao.find(modelId).then(function(m) {
+        var m;
+        this.pending[modelId] = modelDao.find(modelId).then(function(model) {
+          m = model;
           m.validate();
-          if (X.isRegistered(modelId)) return m;
+          return m.arequire(deps);
+        }).then(function() {
```

Comment:
It might be useful to have model.arequire() resolve to the model.

Would make this slightly shorter as you could just get m as a parameter to this promise.

Not a big deal though, but something you could do if you wanted.